plugins/parade-lspcon/fu-parade-lspcon-device.c

```diff
@@ -0,0 +1,832 @@
+				 guint8 register_addr,
+				 guint8 value,
+				 GError **error)
+{
+	guint8 transaction[] = { register_addr, value };
+	return fu_i2c_device_write_full (FU_I2C_DEVICE (self), transaction,
+					 sizeof(transaction), error);
+}
+
+static gboolean
+fu_parade_lspcon_read_register (FuParadeLspconDevice *self,
+				guint8 register_addr,
+				guint8 *value,
+				GError **error)
+{
+	FuI2cDevice *i2c_device = FU_I2C_DEVICE (self);
+	if (!fu_i2c_device_write (i2c_device, register_addr, error))
+		return FALSE;
+	return fu_i2c_device_read (i2c_device, value, error);
+}
+
+/* map the page containing the given address into page 7 */
+static gboolean
+fu_parade_lspcon_map_page (FuParadeLspconDevice *self, guint32 address, GError **error)
+{
+	if (!fu_parade_lspcon_write_register (self,
+					      REG_ADDR_FLASH_ADDR_HI,
+					      address >> 16,
+					      error))
+		return FALSE;
+	return fu_parade_lspcon_write_register (self,
+						REG_ADDR_FLASH_ADDR_LO,
+						address >> 8,
+						error);
+}
+
+/* wait until the specified register masked with mask reads the expected
+ * value, up to 10 seconds */
+static gboolean
+fu_parade_lspcon_poll_register (FuParadeLspconDevice *self,
+				guint8 register_address,
+				guint8 mask,
+				guint8 expected,
+				GError **error)
+{
+	guint8 value;
+	g_autoptr(GTimer) timer = g_timer_new();
+
+	do {
+		if (!fu_parade_lspcon_read_register (self,
+						     register_address,
+						     &value,
+						     error))
+			return FALSE;
+		if ((value & mask) == expected)
+			return TRUE;
+	} while (g_timer_elapsed (timer, NULL) <= 10.0);
+
+	g_set_error (error,
+		     G_IO_ERROR,
+		     G_IO_ERROR_TIMED_OUT,
+		     "register %x did not read %x (mask %x) within 10 seconds: read %x",
+		     register_address, expected, mask, value);
+	return FALSE;
+}
+
+static gboolean
+fu_parade_lspcon_flash_read (FuParadeLspconDevice *self,
+			     guint32 base_address,
+			     guint8 *data,
+			     gsize len,
+			     GError **error)
+{
+	FuDevice *device = FU_DEVICE (self);
+	FuI2cDevice *i2c_device = FU_I2C_DEVICE (self);
+	const gsize target_len = len;
+
+	while (len > 0) {
+		/* page 7 reads always start from the base of the mapped window- we'll
+		 * read the whole page then pull out the parts we care about, using the
+		 * full page everywhere except possibly in the first and last reads */
+		guint8 page_data[256] = { 0x0 };
+		guint8 page_data_start = base_address & 0xFF;
+		gsize page_data_take = MIN ((gssize) len, 256 - page_data_start);
+		g_autoptr(FuParadeLspconI2cAddressGuard) guard = NULL;
+
+		if (!fu_parade_lspcon_map_page (self, base_address, error))
+			return FALSE;
+		guard = fu_parade_lspcon_i2c_address_guard_new (self,
+								I2C_ADDR_PAGE7,
+								error);
+		if (guard == NULL)
+			return FALSE;
+		if (!fu_i2c_device_read_full (i2c_device, page_data, 256, error))
+			return FALSE;
+
+		memcpy (data, page_data + page_data_start, page_data_take);
```

Comment:
Definately use `fu_memcpy_safe` here please. I think it's safe but with the helper I'm not asking the various security teams to think. Then LGTM I think.

Reply:
Done.